roofit/roofitcore/src/TestStatistics/LikelihoodWrapper.cxx

```diff
@@ -0,0 +1,159 @@
+
+void LikelihoodWrapper::synchronizeWithMinimizer(const ROOT::Math::MinimizerOptions & /*options*/) {}
+
+void LikelihoodWrapper::synchronizeParameterSettings(
+   const std::vector<ROOT::Fit::ParameterSettings> & /*parameter_settings*/)
+{
+}
+
+void LikelihoodWrapper::constOptimizeTestStatistic(RooAbsArg::ConstOpCode opcode, bool doAlsoTrackingOpt)
+{
+   likelihood_->constOptimizeTestStatistic(opcode, doAlsoTrackingOpt);
+}
+
+double LikelihoodWrapper::defaultErrorLevel() const
+{
+   return likelihood_->defaultErrorLevel();
+}
+std::string LikelihoodWrapper::GetName() const
+{
+   return likelihood_->GetName();
+}
+std::string LikelihoodWrapper::GetTitle() const
+{
+   return likelihood_->GetTitle();
+}
+
+void LikelihoodWrapper::enableOffsetting(bool flag)
+{
+   do_offset_ = flag;
+   // Clear offset if feature is disabled so that it is recalculated next time it is enabled
+   if (!do_offset_) {
+      offset_ = 0;
+      offset_carry_ = 0;
+   }
+}
+
+void LikelihoodWrapper::setOffsettingMode(OffsettingMode mode)
+{
+   offsetting_mode_ = mode;
+   if (isOffsetting()) {
+      oocoutI(static_cast<RooAbsArg *>(nullptr), Minimization) << "LikelihoodWrapper::setOffsettingMode(" << GetName() << "): changed offsetting mode while offsetting was enabled; resetting offset values" << std::endl;
+      offset_ = 0;
+      offset_carry_ = 0;
+   }
+}
+
+void LikelihoodWrapper::applyOffsetting(double &current_value, double &carry)
+{
+   if (do_offset_) {
+
+      // If no offset is stored enable this feature now
+      if (offset_ == 0 && current_value != 0) {
+         offset_ = current_value;
+         offset_carry_ = carry;
+         if (offsetting_mode_ == OffsettingMode::legacy) {
+            auto sum_likelihood = dynamic_cast<RooSumL*>(likelihood_.get());
+            if (sum_likelihood != nullptr) {
+               double subsidiary_value, subsidiary_carry;
+               std::tie(subsidiary_value, subsidiary_carry) = sum_likelihood->getSubsidiaryValue();
+               // "undo" the addition of the subsidiary value to emulate legacy behavior
+               offset_ -= subsidiary_value;
+               offset_carry_ -= subsidiary_carry;
+               // then add 0 in Kahan summation way to make sure the carry gets taken up into the value if it should be
+               double y = 0 - offset_carry_;
+               double t = offset_ + y;
+               offset_carry_ = (t - offset_) - y;
+               offset_ = t;
+               // also set carry to this value, again to emulate legacy behavior
+               carry = offset_carry_;
+            }
+         }
+         oocoutI(static_cast<RooAbsArg *>(nullptr), Minimization)
+            << "LikelihoodWrapper::applyOffsetting(" << GetName() << "): Likelihood offset now set to " << offset_
+            << std::endl;
+      }
+
+      // Subtract offset
+      // old method:
+//      {
+//         double y = -offset_ - (carry + offset_carry_);
+//         double t = current_value + y;
```

Comment:
Delete

Reply:
I added a TODO for now, because it should be checked whether replacement by KahanSum doesn't affect result values.